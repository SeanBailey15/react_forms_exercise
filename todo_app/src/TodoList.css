.TodoList-h1 {
  text-align: center;
}

.TodoList-form {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  margin: 10px;
}

.TodoList-todos {
  display: flex;
  flex-direction: column;
  align-items: center;
  margin-top: 50px;
}
