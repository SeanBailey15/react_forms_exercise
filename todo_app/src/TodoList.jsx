import React, {useState} from "react";
import {v4 as uuid} from 'uuid'
import Todo from "./Todo";
import NewTodoForm from "./NewTodoForm";
import './TodoList.css'

const TodoList = () => {
    const INITIAL_STATE = []

    const [todos, setTodos] = useState(INITIAL_STATE);

    const addTodo = (newTodo) => {
        setTodos(todos => [...todos, {...newTodo, id: uuid()}])
    }

    const removeTodo = (id) => {
       setTodos(todos.filter((todo) => todo.id !== id))     
    }

    return (
        <div className="TodoList">
            <h1 className="TodoList-h1">Let's Do This!</h1>
            <div className="TodoList-form">
                <NewTodoForm addTodo={addTodo}/>
            </div>
            <div className="TodoList-todos">
                {todos.map(({id, content}) => <Todo key={id} id={id} content={content} removeTodo={removeTodo} />)}
            </div>
        </div>
    )
}

export default TodoList;