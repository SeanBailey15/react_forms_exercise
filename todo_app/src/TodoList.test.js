import React from "react";
import { render, fireEvent, screen, act } from "@testing-library/react";
import TodoList from "./TodoList";

it("renders without crashing", () => {
  act(() => {
    render(<TodoList />);
  });
});

it("matches snapshot", () => {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

it("should add a new todo", async () => {
  act(() => {
    render(<TodoList />);
  });
  const content = screen.getByLabelText("Todo");
  const addButton = screen.queryByText("Add A Task");

  //check if button to remove a todo is present
  expect(screen.queryByText("X")).not.toBeInTheDocument();

  //input todo content, then click to submit
  fireEvent.change(content, { target: { value: "Wash the car" } });
  fireEvent.click(addButton);

  //check that button to remove a todo is now present
  const closeButton = await screen.findByText("X");
  expect(closeButton).toBeInTheDocument();

  //check that todo is present and content is as expected
  const todo = screen.getByTestId("todo");
  expect(todo).toBeInTheDocument();
  expect(todo.textContent).toEqual("Wash the car");
});

it("should remove a todo", async () => {
  act(() => {
    render(<TodoList />);
  });
  const content = screen.getByLabelText("Todo");
  const addButton = screen.queryByText("Add A Task");

  expect(screen.queryByText("X")).not.toBeInTheDocument();

  fireEvent.change(content, { target: { value: "Wash the car" } });
  fireEvent.click(addButton);

  const removeButton = await screen.findByText("X");
  expect(removeButton).toBeInTheDocument();

  const todo = screen.getByTestId("todo");
  expect(todo).toBeInTheDocument();
  expect(todo.textContent).toEqual("Wash the car");

  fireEvent.click(removeButton);

  expect(todo).not.toBeInTheDocument();
  expect(removeButton).not.toBeInTheDocument();
});
