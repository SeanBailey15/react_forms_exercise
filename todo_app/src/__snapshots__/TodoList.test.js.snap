// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TodoList"
  >
    <h1
      class="TodoList-h1"
    >
      Let's Do This!
    </h1>
    <div
      class="TodoList-form"
    >
      <div
        class="NewTodoForm"
      >
        <form
          class="NewTodoForm-form"
        >
          <label
            class="NewTodoForm-label"
            for="content"
          >
            Todo
          </label>
          <input
            class="NewTodoForm-input"
            id="content"
            name="content"
            placeholder="What to do, what to do?"
            type="text"
            value=""
          />
          <button
            class="NewTodoForm-btn"
            type="submit"
          >
            Add A Task
          </button>
        </form>
      </div>
    </div>
    <div
      class="TodoList-todos"
    />
  </div>
</DocumentFragment>
`;
